parachain/pallets/outbound-queue/src/lib.rs

```diff
@@ -344,9 +501,22 @@ pub mod pallet {
 				ticket.message.as_bounded_slice(),
 				AggregateMessageOrigin::Parachain(ticket.origin),
 			);
+			Self::charge_fees(&ticket).map_err(|_| SubmitError::ChargeFeeFailed)?;
```

Comment:
Just to clarify some design goals which I think are important. I don't think `OutboundQueue` should be charging fees in its message processing pipeline, because of the various exceptions and edge cases.

It is better that the actual charging of fees is done in:
1. The control pallet, for permissionless commands such as `CreateAgent`
2. For `AgentExecute` by the the XCM executor, using https://github.com/paritytech/polkadot-sdk/pull/1234 when it is ready. This means we can't make full progress until that is implemented.
  i. Specify the fee (specifically a static base fee) to charge in our [ExportXcm::validate](https://github.com/Snowfork/snowbridge/blob/4f739fde129de640c667dc10eb0a08b701183bdc/parachain/primitives/router/src/outbound/mod.rs#L140) implementation.
 ii. Then the XCM executor will automatically transfer it to the treasury account.

For charging fees in the control pallet, I suggest this approach:

Define a new trait `ChargeFees`, that is implemented by the `OutboundQueue` pallet:

```
trait ChargeFee {
   fn charge_fee(command: Command, from: AccountId) -> Result<V, E>
}
```

Then the control pallet can use this interface to charge fees for various commands depending on whether they are permissionless or not. We should collect all the various fees in a single treasury account controlled by BridgeHub, for example in:

https://github.com/paritytech/polkadot-sdk/pull/1234/files#diff-984ede31f860995ed9e3720b323fce903524085f550b1a6c5a2a128664aa5846R64

Reply:
> charging of fees is done in the control pallet

I still prefer the outbound pallet. The reason is that even we do not charge extra fees for some of the command(e.g. AgentCommand) we still need to provide some estimate api(I may add some runtime rpc for that in outbound pallet) for the source chain to collect fees based upon. So anyway, it's nessessary to take gas cost and reward mechanism into consideration as part of the protocol from the beginning.

> About [#1234](https://github.com/paritytech/polkadot-sdk/pull/1234) the fee manager

Yes, that's the correct way to handle fee. But as commented [here](https://github.com/Snowfork/snowbridge/blob/86c3ffc07f6d64d0078b2bd1a0752fcfbe012d3a/parachain/pallets/outbound-queue/src/lib.rs#L437-L448) this is a workaround before that PR released and I've add `estimate_fee` in https://github.com/Snowfork/snowbridge/blob/86c3ffc07f6d64d0078b2bd1a0752fcfbe012d3a/parachain/primitives/router/src/outbound/mod.rs#L139-L142 for preparation. When that pr got merged we just remove the `charge_fee` logic in outbound pallet.

Reply:
> I still prefer done in outbound pallet. The reason is that even we do not charge extra fees for some of the command(e.g. AgentCommand) we still need to provide some estimate api for the source chain to collect fees based upon(We may add some runtime rpc for that in outbound pallet later). Meanwhile control pallet is just one of the upper layer modules to consume the outbound queue and the charge logic should applies all the same.

It is true that the control pallet consumes the outbound queue, but that does not mean we must push all complexity into the lower level. In fact, it should be the other way around, higher layers add more complexity in an incremental fashion. 

One specific problem with your approach is we have provide information about the agent/origin location to the lower level  (the outbound-queue) so that it can charge fees. In my view, this complicates the `Message` type unnecessarily.

Also we are already using the control pallet for regular governance operations, I think it makes sense to implement `set_outbound_fee_config` extrinsic in the control pallet.

> Yes, that's the correct way to handle fee. But as commented here this is a workaround before that PR released

I don't want us to spend effort on temporary workarounds for what it is a critical aspect of our bridge operation.

I would much rather we only make partial progress, which would mean:
1. Charging fees in control pallet for certain commands
2. Specifying the fee for `AgentExecute` command in our implementation of `ExportXcm::validate`.

Reply:
Furthermore, I also realise now, that if we limit the scope to the audit fix, then we shouldn't need to worry about charging a base fee.

Which means we only need to implement task 1: Charging fees in control pallet for certain commands

Reply:
Revamp for charging fees in control pallet for certain commands
https://github.com/Snowfork/snowbridge/pull/940/commits/448963af70d72d568257a2903cd8129a58a021b3

For task2: Specifying the fee for AgentExecute command in our implementation of ExportXcm::validate. actually it's already covered in current [impl](https://github.com/Snowfork/snowbridge/blob/e8308f8f8dadf9705e4299bc05180f9cf3022344/parachain/primitives/router/src/outbound/mod.rs#L138-L143) so I would prefer not to revert it.